### PR TITLE
escape normalizing Paths

### LINKED ISSUE #9231: 'C/C++:g++.exe build active file' won't complete the build
Issue Type: <b>Bug</b>

I am trying to build the file I'm actively working on. However, when I run the build file task it never finishes. This goes for all of my .cpp files. There a pinsheel that spins showing that the task is still running, and when I try to run task again it says that an instance of the task is still in progress. Closing the broweser and repoening doens't help. Trying a different terminal doesn't help. 

Extension version: 1.10.0
VS Code version: Code 1.66.2 (dfd34e8260c270da74b5c2d86d61aee4b6d56977, 2022-04-11T07:46:01.075Z)
OS version: Windows_NT x64 10.0.22000
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-10210U CPU @ 1.60GHz (8 x 2112)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|7.76GB (1.94GB free)|
|Process Argv|--crash-reporter-id e8fe0db0-a5d4-43b6-aea9-a00cbe79da86|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492cf:30256860
pythonvspyl392:30443607
pythontb:30283811
pythonvspyt551cf:30345471
pythonptprofiler:30281270
vsdfh931:30280409
vshan820:30294714
vstes263:30335439
pythondataviewer:30285071
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593cf:30376535
vsc1dst:30438360
pythonvs932:30410667
wslgetstarted:30449410
pythonvsnew555:30457759
vscscmwlcmt:30465135
cppdebugcf:30475217

```

</details>

### LINKED ISSUE #9241: Build Task does not end - `createDirIfNotExistsSync` fails to create directory due to invalid string quoting
### Environment

- OS and version: MacOS Monterey 12.2.1
- VS Code: 1.66.2
- C/C++ extension: all extensions below
- OS and version of remote machine (if applicable): na
- GDB / LLDB version: 
Issue Type: <b>Bug</b>

VSC version 1.66.2 Mac x86
default build task with clang on Mac stopped working after update.
Proper clang command string generated, terminal opened and command issued there but did not end.
"Building" icon spinning endlessly in info bar at the bottom.
The same command put into terminal (VSC one) will compile super fast successfully.

VS Code version: Code 1.66.2 (dfd34e8260c270da74b5c2d86d61aee4b6d56977, 2022-04-11T07:49:20.994Z)
OS version: Darwin x64 21.3.0
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz (8 x 2200)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|3, 3, 3|
|Memory (System)|16.00GB (1.56GB free)|
|Process Argv|--crash-reporter-id 8ea3aeb5-7459-4aa2-8fe8-57bf4a0e1788|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (31)</summary>

Extension|Author (truncated)|Version
---|---|---
pascal|ale|9.5.0
pascal-formatter|ale|2.6.0
tasks-shell-input|aug|1.7.0
vscode-standardjs|che|1.4.1
doxdocgen|csc|1.4.0
githistory|don|0.6.19
language-forth|ftt|0.0.5
vscode-pull-request-github|Git|0.40.0
r|Iku|2.4.0
linter-gfortran|krv|2.6.2
mads|mir|0.0.5
csharp|ms-|1.24.4
python|ms-|2022.4.1
vscode-pylance|ms-|2022.4.2
jupyter|ms-|2022.3.1000901801
jupyter-keymap|ms-|1.0.0
jupyter-renderers|ms-|1.0.6
remote-containers|ms-|0.231.6
remote-ssh|ms-|0.78.0
remote-ssh-edit|ms-|0.80.0
remote-wsl|ms-|0.66.0
cmake-tools|ms-|1.10.5
cpptools|ms-|1.10.0
cpptools-extension-pack|ms-|1.1.0
vsliveshare|ms-|1.0.5527
vsliveshare-audio|ms-|0.1.91
ossaction|pga|1.0.0
r-debugger|RDe|0.4.7
prolog|reb|0.0.4
vscode-scheme|sjh|0.4.0
html-preview-vscode|tht|0.2.5

(1 theme extensions excluded)

</details>

### Bug Summary and Steps to Reproduce

Bug Summary:

default build task with clang on Mac stopped working after update.
Proper clang command string generated, terminal opened and command issued there but did not end.
"Building" icon spinning endlessly in info bar at the bottom.
The same command put into terminal (VSC one) will compile super fast successfully.

Steps to reproduce:
1. In this environment (VSCode terminal opened by default is zsh)
2. With this config: tasks.json:
```
{
	"version": "2.0.0",
	"tasks": [
		{
			"type": "cppbuild",
			"label": "C/C++: clang kompiluj aktywny plik",
			"command": "/usr/bin/clang",
			"args": [
				"-fdiagnostics-color=always",
				"-g",
				"${file}",
				"-o",
				"${fileDirname}/${fileBasenameNoExtension}"
			],
			"options": {
				"cwd": "${fileDirname}"
			},
			"problemMatcher": [
				"$gcc"
			],
			"group": {
				"kind": "build",
				"isDefault": true
			},
			"detail": "kompilator: /usr/bin/clang"
		}
	]
}

```
3. Do Shift-Command-B
4. See error - Never Ending Compilation (I guess it didn't really started except invoking command):
Zrzut ekranu 2022-04-26 o 22 40 12

### Debugger Configurations

```shell
{
	"version": "2.0.0",
	"tasks": [
		{
			"type": "cppbuild",
			"label": "C/C++: clang kompiluj aktywny plik",
			"command": "/usr/bin/clang",
			"args": [
				"-fdiagnostics-color=always",
				"-g",
				"${file}",
				"-o",
				"${fileDirname}/${fileBasenameNoExtension}"
			],
			"options": {
				"cwd": "${fileDirname}"
			},
			"problemMatcher": [
				"$gcc"
			],
			"group": {
				"kind": "build",
				"isDefault": true
			},
			"detail": "kompilator: /usr/bin/clang"
		}
	]
}
```

### Debugger Logs

```shell
No output from Tasks
```

### Other Extensions

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by elahehrashedi**
bug fix: https://github.com/microsoft/vscode-cpptools/issues/9231
bug fix: https://github.com/microsoft/vscode-cpptools/issues/9241

### CODE REVIEW

File: Extension/src/LanguageServer/cppBuildTaskProvider.ts [hunk 1]
```diff
@@ -361,6 +361,11 @@ class CustomBuildTaskTerminal implements Pseudoterminal {
         // Do build.
         let command: string = util.resolveVariables(this.command);
         let activeCommand: string = command;
+
+        // Create the exe folder path if it doesn't exists.
+        const exePath: string | undefined = util.resolveVariables(util.findExePathInArgs(this.args));
+        util.createDirIfNotExistsSync(exePath);
+
         this.args.forEach((value, index) => {
```
Comment (Colengms) [inline Extension/src/LanguageServer/cppBuildTaskProvider.ts:370]: Something I just noticed/realized about this function: We should not be adding spaces or quoting to arguments in an argument array.  It's unnecessary.  We are passing these arguments to `spawn`, which accepts them without quoting/escaping, unless we specify `shell` in the options argument, which we do not need to do.  The adding of `shell` to the options arg should be removed, and we should stop modifying the arguments to add quoting, as it just adds complexity to what could be a simple operation.  In cases where it's necessary to provide a full command line with (quoted) arguments in it, our implementation of normalizeArg is wrong, and should instead be made equivilent to `quote_argument` in GlueWin.cpp and GlueXnix.cpp on the native side.

**Review by Colengms — CHANGES_REQUESTED** [review]

File: Extension/src/LanguageServer/cppBuildTaskProvider.ts (hunk 1 above)
Comment (Colengms) [reply Extension/src/LanguageServer/cppBuildTaskProvider.ts:370]: https://nodejs.org/api/child_process.html#child_processspawncommand-args-options

Comment (Colengms) [reply Extension/src/LanguageServer/cppBuildTaskProvider.ts:370]: Or, if the issue is that we need to invoke using the shell for other reasons (shell script tasks?), then the issue is only that normalizeArgs needs to be updated to match `quote_argument` in GlueWin.cpp and GlueXnix.cpp on the native side.

Comment (Colengms) [reply Extension/src/LanguageServer/cppBuildTaskProvider.ts:370]: As discussed, we will update `normalizeArgs` in a subsequent PR.

**Review by Colengms — APPROVED** [review]

File: Extension/src/LanguageServer/cppBuildTaskProvider.ts [hunk 2]
```diff
@@ -361,6 +361,11 @@ class CustomBuildTaskTerminal implements Pseudoterminal {
         // Do build.
         let command: string = util.resolveVariables(this.command);
         let activeCommand: string = command;
+
+        // Create the exe folder path if it doesn't exists.
```
Comment (jogo-) [inline Extension/src/LanguageServer/cppBuildTaskProvider.ts:366]: Typo: exist**s** -> exist

Comment (sean-mcmanus) [reply Extension/src/LanguageServer/cppBuildTaskProvider.ts:366]: Thanks -- I've added the fix to https://github.com/microsoft/vscode-cpptools/pull/9270

**Pull request merged**